packages/core/src/transaction-promise.ts

```diff
@@ -0,0 +1,193 @@
+   * @param {function()} onClose - Function to be called when transaction is committed or rolled back.
+   * @param {function(bookmarks: Bookmarks)} onBookmarks callback invoked when new bookmark is produced.
+   * * @param {function()} onConnection - Function to be called when a connection is obtained to ensure the conneciton
+   * is not yet released.
+   * @param {boolean} reactive whether this transaction generates reactive streams
+   * @param {number} fetchSize - the record fetch size in each pulling batch.
+   * @param {string} impersonatedUser - The name of the user which should be impersonated for the duration of the session.
+   */
+  constructor({
+    connectionHolder,
+    onClose,
+    onBookmarks,
+    onConnection,
+    reactive,
+    fetchSize,
+    impersonatedUser,
+    highRecordWatermark,
+    lowRecordWatermark
+  }: {
+    connectionHolder: ConnectionHolder
+    onClose: () => void
+    onBookmarks: (bookmarks: Bookmarks) => void
+    onConnection: () => void
+    reactive: boolean
+    fetchSize: number
+    impersonatedUser?: string,
+    highRecordWatermark: number,
+    lowRecordWatermark: number
+  }) {
+    super({
+      connectionHolder,
+      onClose,
+      onBookmarks,
+      onConnection,
+      reactive,
+      fetchSize,
+      impersonatedUser,
+      highRecordWatermark,
+      lowRecordWatermark
+    })
+  }
+
+  /**
+   * Waits for the begin to complete.
+   *
+   * @param {function(transaction: Transaction)} onFulfilled - function to be called when finished.
+   * @param {function(error: {message:string, code:string})} onRejected - function to be called upon errors.
+   * @return {Promise} promise.
+   */
+  then<TResult1 = Transaction, TResult2 = never>(
+    onfulfilled?:
+      ((value: Transaction) => TResult1 | PromiseLike<TResult1>)
+      | null,
+    onrejected?:
+      ((reason: any) => TResult2 | PromiseLike<TResult2>)
+      | null
+  ): Promise<TResult1 | TResult2> {
+    return this._getOrCreateBeginPromise().then(onfulfilled, onrejected);
+  }
+
+  /**
+   * Catch errors when using promises.
+   *
+   * @param {function(error: Neo4jError)} onRejected - Function to be called upon errors.
+   * @return {Promise} promise.
+   */
+  catch<TResult = never>(onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | null): Promise<any> {
+    return this._getOrCreateBeginPromise().catch(onrejected);
+  }
+
+  /**
+   * Called when finally the begin is done
+   *
+   * @param {function()|null} onfinally - function when the promise finished
+   * @return {Promise} promise.
+   */
+  finally(onfinally?: (() => void) | null): Promise<Transaction> {
+    return this._getOrCreateBeginPromise().finally(onfinally);
+  }
+
+  private _getOrCreateBeginPromise(): Promise<Transaction> {
+    if (!this._beginPromise) {
+      this._beginPromise = new Promise((resolve, reject) => {
+        this._resolve = resolve;
+        this._reject = reject;
+        if (this._beginError) {
```

Comment:
how can this condition be truthy when the promise has not been initialized yet? I have trouble following the flow here

Reply:
That's fun. As shown bellow, the begin method already subscribe to the begin observer and keep track of the errors and any promises are created. So, if the user took sufficient time before await in the transaction, the `beginError` or `beginMetadata` could be already created. So the promise return by `_getOrCreateBeginPromise` will be already solved.

```
_begin(bookmarks: string | Bookmarks | string[], txConfig: TxConfig): void {
    return super._begin(bookmarks, txConfig, {
      onError: this._onBeginError.bind(this),
      onComplete: this._onBeginMetadata.bind(this)
    });
  }
```

Why do I not create the promise as soon as I create the TxPromise object? Simple, if the promise get reject without anyone catching it,  It will have `UncaughtPromiseException` in newer versions of Nodejs. So, I have to postpone the internal promise creation until the moment the user tries to resolve `TransactionPromise` and keep the begin events in memory for when i need it.

Reply:
Yes, it is a misuse. But the TransactionPromise should behave as the Transaction, so the user could choose for not trying to resolve it without any issue. 

But if the user call `then` or `finally` without have a catch in the chain, they will still have the `UncaughtPromiseException` as expected.